File: packages/asdl-slots/src/asdl_slots/lifecycle/checkout.py
"""Branch checkout into the slot lifecycle pool."""

from __future__ import annotations

import subprocess

from asdl_core.git.types import WorktreeOccupancy
from asdl_slots.checkout_planning import (
    AssignToSlot,
    BranchInMainWorktree,
    BranchInUse,
    CheckoutCurrentWorktreeBranch,
    CurrentCheckoutPlan,
    CurrentWorktreeRedirect,
    DetachCurrentWorktreeHead,
    PoolFull,
    ReuseAssignment,
    plan_checkout,
    plan_current_checkout,
)
from asdl_slots.context import SlotsCliContext
from asdl_slots.errors import (
    DetachedHeadError,
    DirtyCurrentWorktreeError,
    SlotAllocationError,
)
from asdl_slots.inventory import SlotRecord, build_slot_inventory
from asdl_slots.lifecycle.operation_state import operation_recovery_sentence
from asdl_slots.lifecycle.outcomes import SlotCheckoutOutcome, SlotLifecycleFailure
from asdl_slots.repo_context import ensure_slots_metadata_dir

ExecutableCheckoutPlan = ReuseAssignment | BranchInMainWorktree | AssignToSlot


def _assigned_detail(record: SlotRecord) -> str:
    suffix = ""
    if record.operation is not None:
        suffix = f" ({record.operation} in progress)"
    return f"  {record.slot_name} -> {record.branch}{suffix}"


def checkout_branch(
    slots_ctx: SlotsCliContext,
    branch_name: str,
    *,
    new_branch: bool,
    base: str | None,
) -> SlotCheckoutOutcome | SlotLifecycleFailure:
    ensure_slots_metadata_dir(slots_ctx.repo, slots_ctx.storage)

    branch_exists = slots_ctx.git.branch_exists(branch_name)
    created_branch = False
    if new_branch:
        if branch_exists:
            return SlotLifecycleFailure(
                error_type="branch_exists",
                message=(
                    f"Branch '{branch_name}' already exists. "
                    "Drop -b to check out the existing branch."
                ),
            )
        if base is not None and not slots_ctx.git.branch_exists(base):
            return SlotLifecycleFailure(
                error_type="base_missing",
                message=f"Base branch '{base}' does not exist.",
            )
        slots_ctx.git.create_branch(
            branch_name,
            base if base is not None else "HEAD",
            force=False,
        )
        created_branch = True
    elif not branch_exists:
        return SlotLifecycleFailure(
            error_type="branch_missing",
            message=(
                f"Branch '{branch_name}' does not exist. Pass -b/--new to create it from HEAD."
            ),
        )

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    plan = plan_checkout(inventory, slots_ctx.git, branch_name)
    if isinstance(plan, PoolFull):
        return _pool_full_failure(plan)
    if isinstance(plan, BranchInUse):
        return _branch_in_use_failure(plan.occupancy)
    return _execute_plan(
        plan,
        slots_ctx=slots_ctx,
        branch_name=branch_name,
        created_branch=created_branch,
        current_wt_note=None,
    )


def checkout_current(slots_ctx: SlotsCliContext) -> SlotCheckoutOutcome | SlotLifecycleFailure:
    ensure_slots_metadata_dir(slots_ctx.repo, slots_ctx.storage)

    try:
        current_plan = plan_current_checkout(
            slots_ctx.git,
            cwd=slots_ctx.repo.root,
            main_repo_root=slots_ctx.repo.main_repo_root,
        )
    except SlotAllocationError as exc:
        return SlotLifecycleFailure(
            error_type="slot_allocation_error",
            message=str(exc),
        )

    if isinstance(current_plan, DetachedHeadError):
        return SlotLifecycleFailure(
            error_type="detached_head",
            message=(
                f"HEAD at {current_plan.cwd} is detached. Check out a branch "
                "before running `slot checkout --current`."
            ),
        )
    if isinstance(current_plan, DirtyCurrentWorktreeError):
        return SlotLifecycleFailure(
            error_type="dirty_worktree",
            message=(
                f"Current worktree at {current_plan.cwd} has uncommitted changes. "
                "Commit or stash before running `slot checkout --current`."
            ),
        )

    assert isinstance(current_plan, CurrentCheckoutPlan)
    if isinstance(current_plan.plan, PoolFull):
        return _pool_full_failure(current_plan.plan)
    if isinstance(current_plan.plan, BranchInUse):
        return _branch_in_use_failure(current_plan.plan.occupancy)

    redirect_result = _execute_current_wt_redirect(
        current_plan.current_wt_redirect,
        slots_ctx=slots_ctx,
    )
    if isinstance(redirect_result, SlotLifecycleFailure):
        return redirect_result

    return _execute_plan(
        current_plan.plan,
        slots_ctx=slots_ctx,
        branch_name=current_plan.branch_name,
        created_branch=False,
        current_wt_note=redirect_result,
    )


def _execute_current_wt_redirect(
    redirect: CurrentWorktreeRedirect | None,
    *,
    slots_ctx: SlotsCliContext,
) -> str | SlotLifecycleFailure | None:
    if redirect is None:
        return None
    if isinstance(redirect, CheckoutCurrentWorktreeBranch):
        failure = slots_ctx.git.checkout_branch(slots_ctx.repo.root, redirect.branch_name)
        if failure is not None:
            return SlotLifecycleFailure(
                error_type="slot_allocation_error",
                message=(
                    f"Failed to check out '{redirect.branch_name}' in "
                    f"{slots_ctx.repo.root}: {failure.message}"
                ),
            )
        return redirect.note

    if isinstance(redirect, DetachCurrentWorktreeHead):
        try:
            slots_ctx.git.detach_head(slots_ctx.repo.root, redirect.ref)
        except subprocess.CalledProcessError as exc:
            stderr = exc.stderr.strip() if isinstance(exc.stderr, str) and exc.stderr else str(exc)
            return SlotLifecycleFailure(
                error_type="slot_allocation_error",
                message=(
                    f"Failed to detach current worktree at {slots_ctx.repo.root} "
                    f"to {redirect.ref}: {stderr}"
                ),
            )
        return redirect.note

    return None


def _pool_full_failure(outcome: PoolFull) -> SlotLifecycleFailure:
    if outcome.assigned:
        details = "\n".join(_assigned_detail(record) for record in outcome.assigned)
        message = (
            f"Pool is full. Currently assigned:\n{details}\n"
            "Free a slot before checking out a new branch."
        )
    else:
        message = (
            "Pool is full (no slots available). "
            "Run `slot init` or `slot resize` before checking out a new branch."
        )
    return SlotLifecycleFailure(error_type="pool_full", message=message)


def _branch_in_use_failure(occupancy: WorktreeOccupancy) -> SlotLifecycleFailure:
    if occupancy.operation in ("rebase", "bisect"):
        message = (
            f"Branch '{occupancy.branch}' has a {occupancy.operation} in progress at "
            f"{occupancy.path}. {operation_recovery_sentence(occupancy.operation)}, then retry."
        )
    else:
        message = f"Branch '{occupancy.branch}' is already checked out at {occupancy.path}."
    return SlotLifecycleFailure(error_type="branch_in_use", message=message)


def _execute_plan(
    plan: ExecutableCheckoutPlan,
    *,
    slots_ctx: SlotsCliContext,
    branch_name: str,
    created_branch: bool,
    current_wt_note: str | None,
) -> SlotCheckoutOutcome | SlotLifecycleFailure:
    if isinstance(plan, ReuseAssignment):
        return SlotCheckoutOutcome(
            slot_name=plan.record.slot_name,
            branch_name=branch_name,
            worktree_path=plan.record.path,
            already_assigned=True,
            created_branch=created_branch,
            current_wt_note=current_wt_note,
        )
    if isinstance(plan, BranchInMainWorktree):
        return SlotCheckoutOutcome(
            slot_name="",
            branch_name=branch_name,
            worktree_path=plan.main_path,
            already_assigned=True,
            created_branch=created_branch,
            current_wt_note=current_wt_note,
        )

    failure = slots_ctx.git.checkout_branch(plan.record.path, branch_name)
    if failure is not None:
        return SlotLifecycleFailure(
            error_type="checkout_failed",
            message=(
                f"Failed to check out '{branch_name}' into {plan.record.slot_name}: "
                f"{failure.message}"
            ),
        )
    return SlotCheckoutOutcome(
        slot_name=plan.record.slot_name,
        branch_name=branch_name,
        worktree_path=plan.record.path,
        already_assigned=False,
        created_branch=created_branch,
        current_wt_note=current_wt_note,
    )

File: ts/packages/slot/src/lifecycle/checkout.ts
import type { SlotCliContext } from "../context.ts";
import { buildSlotInventory } from "../inventory.ts";
import { planCheckout, planCurrentCheckout, type CheckoutPlan } from "../planning.ts";
import { ensureSlotsMetadataDir } from "../repo-context.ts";
import { branchInUseFailure, poolFullFailure, type LifecycleResult, type SlotLifecycleFailure } from "./common.ts";
import { executeCurrentWorktreeRedirect } from "./current-worktree-redirect.ts";

export interface SlotCheckoutOutcome {
	slot_name: string;
	branch_name: string;
	worktree_path: string;
	already_assigned: boolean;
	created_branch: boolean;
	current_wt_note: string | null;
}

export type SlotCheckoutResult = LifecycleResult<SlotCheckoutOutcome>;

type ExecutableCheckoutPlan = Extract<CheckoutPlan, { type: "reuse_assignment" | "branch_in_main_worktree" | "assign_to_slot" }>;

export async function checkoutBranch(ctx: SlotCliContext, branchName: string, options: { newBranch: boolean; base: string | null }): Promise<SlotCheckoutResult> {
	if (ctx.repo.type !== "repo") return { type: "failure", failure: { error_type: ctx.repo.errorType, message: ctx.repo.message } };
	await ensureSlotsMetadataDir(ctx.repo, ctx.storage);

	const branchExists = await ctx.git.branchExists(branchName);
	let createdBranch = false;
	if (options.newBranch) {
		if (branchExists) return failure("branch_exists", `Branch '${branchName}' already exists. Drop -b to check out the existing branch.`);
		if (options.base !== null && !(await ctx.git.branchExists(options.base))) return failure("base_missing", `Base branch '${options.base}' does not exist.`);
		const createFailure = await ctx.git.createBranch(branchName, options.base ?? "HEAD", { force: false });
		if (createFailure !== null) return failure("branch_create_failed", `Failed to create branch '${branchName}': ${createFailure.message}`);
		createdBranch = true;
	} else if (!branchExists) {
		return failure("branch_missing", `Branch '${branchName}' does not exist. Pass -b/--new to create it from HEAD.`);
	}

	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	const plan = await planCheckout(inventory, ctx.git, branchName);
	if (plan.type === "pool_full") return { type: "failure", failure: poolFullFailure(plan.assigned, { action: "checking out a new branch" }) };
	if (plan.type === "branch_in_use") return { type: "failure", failure: branchInUseFailure(plan.occupancy) };
	return await executeCheckoutPlan(plan, ctx, { branchName, createdBranch, currentWtNote: null });
}

export async function checkoutCurrent(ctx: SlotCliContext): Promise<SlotCheckoutResult> {
	if (ctx.repo.type !== "repo") return { type: "failure", failure: { error_type: ctx.repo.errorType, message: ctx.repo.message } };
	await ensureSlotsMetadataDir(ctx.repo, ctx.storage);

	const currentPlan = await planCurrentCheckout(ctx.git, { cwd: ctx.repo.root, mainRepoRoot: ctx.repo.mainRepoRoot });
	if (currentPlan.type === "allocation_failure") return failure("slot_allocation_error", currentPlan.message);
	if (currentPlan.type === "detached_head") return failure("detached_head", `HEAD at ${currentPlan.cwd} is detached. Check out a branch before running \`slot checkout --current\`.`);
	if (currentPlan.type === "dirty_worktree") return failure("dirty_worktree", `Current worktree at ${currentPlan.cwd} has uncommitted changes. Commit or stash before running \`slot checkout --current\`.`);
	if (currentPlan.plan.type === "pool_full") return { type: "failure", failure: poolFullFailure(currentPlan.plan.assigned, { action: "checking out a new branch" }) };
	if (currentPlan.plan.type === "branch_in_use") return { type: "failure", failure: branchInUseFailure(currentPlan.plan.occupancy) };
	if (currentPlan.redirect !== null) {
		const redirectFailure = await executeCurrentWorktreeRedirect(currentPlan.redirect, ctx);
		if (redirectFailure !== null) return { type: "failure", failure: redirectFailure };
	}
	return await executeCheckoutPlan(currentPlan.plan, ctx, { branchName: currentPlan.branchName, createdBranch: false, currentWtNote: currentPlan.currentWtNote });
}

async function executeCheckoutPlan(plan: ExecutableCheckoutPlan, ctx: SlotCliContext, options: { branchName: string; createdBranch: boolean; currentWtNote: string | null }): Promise<SlotCheckoutResult> {
	if (plan.type === "reuse_assignment") {
		return ok({ slot_name: plan.record.slotName, branch_name: options.branchName, worktree_path: plan.record.path, already_assigned: true, created_branch: options.createdBranch, current_wt_note: options.currentWtNote });
	}
	if (plan.type === "branch_in_main_worktree") {
		return ok({ slot_name: "", branch_name: options.branchName, worktree_path: plan.mainPath, already_assigned: true, created_branch: options.createdBranch, current_wt_note: options.currentWtNote });
	}
	const checkoutFailure = await ctx.git.checkoutBranch(plan.record.path, options.branchName);
	if (checkoutFailure !== null) return failure("checkout_failed", `Failed to check out '${options.branchName}' into ${plan.record.slotName}: ${checkoutFailure.message}`);
	return ok({ slot_name: plan.record.slotName, branch_name: options.branchName, worktree_path: plan.record.path, already_assigned: false, created_branch: options.createdBranch, current_wt_note: options.currentWtNote });
}

function ok(outcome: SlotCheckoutOutcome): SlotCheckoutResult {
	return { type: "ok", outcome };
}

function failure(errorType: string, message: string): SlotCheckoutResult {
	const lifecycleFailure: SlotLifecycleFailure = { error_type: errorType, message };
	return { type: "failure", failure: lifecycleFailure };
}
